pkg/sfu/connectionquality/scorer.go

```diff
@@ -0,0 +1,443 @@
+	windows []*windowScore
+
+	mutedAt   time.Time
+	unmutedAt time.Time
+
+	maxPPS float64
+
+	transitions []layerTransition
+}
+
+func newQualityScorer(params qualityScorerParams) *qualityScorer {
+	return &qualityScorer{
+		params: params,
+		score:  maxScore,
+		state:  qualityScorerStateStable,
+	}
+}
+
+func (q *qualityScorer) Start(at time.Time) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.lastUpdateAt = at
+}
+
+func (q *qualityScorer) UpdateMute(isMuted bool, at time.Time) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	if isMuted {
+		q.mutedAt = at
+
+		// stable when muted
+		q.state = qualityScorerStateStable
+		q.windows = q.windows[:0]
+		q.score = maxScore
+	} else {
+		q.unmutedAt = at
+	}
+}
+
+func (q *qualityScorer) AddTransition(bitrate int64, at time.Time) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.transitions = append(q.transitions, layerTransition{
+		startedAt: at,
+		bitrate:   bitrate,
+	})
+}
+
+func (q *qualityScorer) Update(stat *windowStat, at time.Time) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	// nothing to do when muted or not unmuted for long enough
+	// NOTE: it is possible that unmute -> mute -> unmute transition happens in the
+	//       same analysis window. On a transition to mute, state immediately moves
+	//       to stable and quality EXCELLENT for responsiveness. On an unmute, the
+	//       entire window data is considered (as long as enough time has passed since
+	//       unmute) include the data before mute.
+	if q.isMuted() || !q.isUnmutedEnough(at) {
+		q.lastUpdateAt = at
+		return
+	}
+
+	var ws *windowScore
+	if stat == nil {
+		ws = newWindowScoreWithScore(&windowStat{
+			startedAt: q.lastUpdateAt,
+			duration:  at.Sub(q.lastUpdateAt),
+		}, poorScore)
+	} else {
+		if stat.packetsExpected == 0 {
+			ws = newWindowScoreWithScore(stat, poorScore)
+		} else {
+			wsPacket := newWindowScorePacket(stat, q.getPacketLossWeight(stat))
+			wsByte := newWindowScoreByte(stat, q.getExpectedBitsAndUpdateTransitions(at))
+			if wsPacket.getScore() < wsByte.getScore() {
+				ws = wsPacket
+			} else {
+				ws = wsByte
+			}
+		}
+	}
+	score := ws.getScore()
+	cq := scoreToConnectionQuality(score)
+
+	q.lastUpdateAt = at
+
+	// transition to start of recovering on any quality drop
+	// WARNING NOTE: comparing protobuf enum values directly (livekit.ConnectionQuality)
+	if scoreToConnectionQuality(q.score) > cq {
+		q.windows = []*windowScore{ws}
+		q.state = qualityScorerStateRecovering
+		q.score = score
+		return
+	}
+
+	// if stable and quality continues to be EXCELLENT, stay there.
+	if q.state == qualityScorerStateStable && cq == livekit.ConnectionQuality_EXCELLENT {
+		q.score = score
+		return
+	}
+
+	// when recovering, look at a longer window
+	q.windows = append(q.windows, ws)
+	if !q.prune(at) {
+		// minimum recovery duration not satisfied, hold at current quality
+		return
+	}
+
+	// take median of scores in a longer window to prevent quality reporting oscillations
```

Comment:
This is the hysteresis. I am not too happy about responsiveness of this. Have set it to 25 seconds of wait if quality drops to `POOR`, i. e. we will take median quality 25 seconds after `POOR` starts. And that wait is 15 seconds for `GOOD`. So, if things are transitioning from `POOR` -> `GOOD` -> `EXCELLENT`, this could take 40 seconds. That might be fine, but there should be better ways. Let me know if you can think of something.

Reason for 25 and 15 seconds is that the analysis window is 5 seconds and wanted to have enough windows to take median.

Reply:
seems fine to me especially for first pass. if it feels too slow reacting, we can come back to it.

Reply:
if this flapped too much the ui updates might be distracting so having a bit of delay might be better user experience

Reply:
yeah, that's the goal. Just don't want it to be too slow 😄